### PR TITLE
Grafana Loki destination

### PR DESCRIPTION
**Pull request by MrAnno**
This PR implements the gRPC-based `loki()` destination.

The message format is the same as in their documented HTTP endpoint:
https://grafana.com/docs/loki/latest/reference/api/#push-log-entries-to-loki

Example config:
```
loki(
    url("localhost:9096")
    # timestamp(msg)
    labels(
        "app" => "$PROGRAM",
        "host" => "$HOST",
    )

    workers(16)
    batch-timeout(10000)
    batch-lines(1000)
);
```

Depends on #4630

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by MrAnno** [comment]
@kira-syslogng do stresstest

**Comment by kira-syslogng** [comment]
Kira-stress-test: Build SUCCESS

**Comment by alltilla** [comment]
LGTM so far :+1:

File: modules/grpc/loki/loki-dest.cpp [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+  if (!this->message)
+    {
+      this->message = log_template_new(cfg, NULL);
+      log_template_compile(this->message, DEFAULT_MESSAGE_TEMPLATE, NULL);
+    }
+
+  log_template_options_init(&this->template_options, cfg);
+
+  LogTemplate *worker_partition_key = log_template_new(cfg, NULL);
```
Comment (alltilla) [inline modules/grpc/loki/loki-dest.cpp:78]: There are some configurations, which do not work right now, can you please suggest, whether they are valid use-cases, and what should be the expected behavior for them?

1. No labels. Currently we get the error: `worker-partition-key() should not be literal string`
2. Static labels (`labels("foo" => "bar"))`). Currently we get the same error as in the previous point.

Maybe no labels is illogical, and we should check it and throw an error for the user, to set labels.

However setting static labels might be a valid thing to do. In this case I think we could warn the user if they have multiple `workers()` set, that we will only use 1, and we should not use the worker partition key logic of the threaded dest at all.

File: modules/grpc/loki/loki-dest.cpp [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+using syslogng::grpc::loki::DestinationDriver;
+
+struct _LokiDestDriver
+{
+  LogThreadedDestDriver super;
+  DestinationDriver *cpp;
+};
+
+DestinationDriver::DestinationDriver(LokiDestDriver *s)
+  : super(s), url("localhost:9096"), timestamp(LM_TS_PROCESSED),
```
Comment (alltilla) [inline modules/grpc/loki/loki-dest.cpp:47]: To me the default port was 9095, and the docs says this, too: https://grafana.com/docs/loki/latest/configure/

```
# gRPC server listen port.
# CLI flag: -server.grpc-listen-port
[grpc_listen_port: <int> | default = 9095]
```

Am I missing something?

File: modules/grpc/loki/loki-worker.cpp [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+  if (this->super->super.batch_size == 0)
+    return LTR_SUCCESS;
+
+  LogThreadedResult result;
+  logproto::PushResponse response{};
+
+  ::grpc::ClientContext ctx;
+  ::grpc::Status status = this->stub->Push(&ctx, this->current_batch, &response);
+
+  if (!status.ok())
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/grpc/loki/loki-worker.cpp:225]: I couldn't find any proper documentation of the gRPC APIs.
I think we could do the mentioned mapping by checking their code, maybe in a next iteration.

File: modules/grpc/loki/loki-dest.cpp (hunk 1 above)
Comment (MrAnno) [reply modules/grpc/loki/loki-dest.cpp:78]: I think both cases are valid and in those cases, we should just use the default round-robin worker indexing.

File: modules/grpc/loki/loki-dest.cpp (hunk 2 above)
Comment (MrAnno) [reply modules/grpc/loki/loki-dest.cpp:47]: You are right. There are signs of 9096 in older(?) configs. Fixed.

**Review by alltilla — APPROVED** [review]

**Pull request merged**